File: 백준/Silver/14501. 퇴사/퇴사.py
import sys
from collections import deque
input = sys.stdin.readline

n = int(input())
calender = [[] for _ in range(n+1)]
for i in range(1,n+1):
    t, p = map(int,input().split())
    calender[i] = [t,p]

max_cost = 0
def bfs(start_day,st,sp):
    global max_cost
    queue = deque()
    queue.append((start_day,sp))

    while queue:
        day, profit = queue.popleft()
        max_cost = max(max_cost, profit)
        if day > n:
            continue
        if day <= n:
            t, p = calender[day]

            if day + t <= n + 1:
                queue.append((day+t, profit + p))

            queue.append((day+1,profit))


for i in range(1,n+1):
    t, p = calender[i]
    bfs(i,0,0)

print(max_cost)

File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.*;

class Main {
  static int answer;
  static int N;
  static int[][] day;
   public static void main(String[] args) throws IOException {
      BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

      N = Integer.parseInt(br.readLine());

      day = new int[N][2];

      for (int i = 0; i < N; i++) {
        String[] st = br.readLine().split(" ");

        int t = Integer.parseInt(st[0]);
        int p = Integer.parseInt(st[1]);

        day[i][0] = t;
        day[i][1] = p;
      }

      int[] dp = new int[N+1];
      for (int i = 0; i < N; i++) {
        int end = i + day[i][0];

        // 상담을 하지 않는 경우
        dp[i+1] = Math.max(dp[i+1], dp[i]);

        // 상담을 하는 경우
        if (end <= N) {
          dp[end] = Math.max(dp[end], day[i][1] + dp[i]);
        }
      }

      System.out.println(dp[N]);

   }

}